pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -345,6 +346,15 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
       requestStatistics.setFanoutType(RequestStatistics.FanoutType.REALTIME);
     }
 
+    // Check if response can be send without server query evaluation.
+    if (isResponsePossible(offlineBrokerRequest) && isResponsePossible(realtimeBrokerRequest)) {
```

Comment:
Set `offlineBrokerRequest` or `realtimeBrokerRequest` to `null` if only one of them is always false

Reply:
Renamed to `isFilterAlwaysFalse` and also added 'isFilterAlwaysTrue' to take care of all cases including setting `offlineBrokerRequest` or `realtimeBrokerRequest` to null if needed and dropping filter if filter will always evaluate to TRUE.